servicetalk-client-api/src/main/java/io/servicetalk/client/api/LoadBalancerFactory.java

```diff
@@ -63,8 +68,8 @@
      */
     default <T extends C> LoadBalancer<T> newLoadBalancer(
             String targetResource,
-            Publisher<? extends ServiceDiscovererEvent<ResolvedAddress>> eventPublisher,
+            Publisher<? extends Collection<? extends ServiceDiscovererEvent<ResolvedAddress>>> eventPublisher,
```

Comment:
Do we need `? extends` for `Collection`? `discovery` method returns `Collection`:
https://github.com/apple/servicetalk/blob/f9119a7910031986750586cae118a572b4f11f1e/servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscoverer.java#L52

I can see the value of giving more flexibility to express a more specific `Collection` type, but only if `ServiceDiscoverer` API also aligned. Not sure if we have to go that path though. WDYT?

Reply:
Consider the following example:

```java
    private static <T> Publisher<? extends T> extract(Publisher<Collection<? extends T>> pub) {
        return pub.flatMapConcatIterable(Function.identity());
    }
    
    final Publisher<Collection<String>> from = Publisher.from(List.of("hello"), List.of("world"));
    extract(from).toFuture().get(); // failure: reason: Incompatible equality constraint: ? extends T and String
```

vs

```java
    private static <T> Publisher<? extends T> extract(Publisher<? extends Collection<? extends T>> pub) {
        return pub.flatMapConcatIterable(Function.identity());
    }
    
    final Publisher<Collection<String>> from = Publisher.from(List.of("hello"), List.of("world"));
    extract(from).toFuture().get(); // compiles fine
```

Full output of the compiler:

```
error: method extract in class (...) cannot be applied to given types;
        extract(from).toFuture().get();
        ^
  required: Publisher<Collection<? extends T>>
  found: Publisher<Collection<String>>
  reason: cannot infer type-variable(s) T
    (argument mismatch; Publisher<Collection<String>> cannot be converted to Publisher<Collection<? extends T>>)
  where T is a type-variable:
    T extends Object declared in method <T>extract(Publisher<Collection<? extends T>>)
```

The reason is that `Publisher<Collection<String>` is not a child of `Publisher<Collection<? extends String>`, but is a child of `Publisher<? extends Collection<? extends String>>` in the same way that `Collection<String>` is not a child of `Collection<Object>` but is a child of  `Collection<? extends Object>`:

```java
Collection<Object> l = List.<String>of("hello"); // doesn't compile
Collection<? extends Object> l2 = List.<String>of("world"); // compiles fine
```

Reply:
Thanks for explaining, sgtm!